### ISSUE TITLE
Type class implementation

### ISSUE THREAD

**Issue by krame505** [melt-umn/silver#366]
Some more thoughts on implementing type classes, separated out from design discussion in #364.

Type classes and instances would be represented in seperate environment namespaces, similarly to how attributes and occurrences are represented in the environment.

Every type class would be translated as an abstract class in Java with an abstract "get" method for every member, and type class instances would be concrete sub-classes of the corresponding abstract class.  For example,
```haskell
class Monoid m {
  mempty :: m;
  mappend :: (m ::= m m);
}

instance Monoid Maybe<a> {
  mempty = nothing();
  mappend = orElse;
}
```
would translate as
```java
abstract class CMonoid<m> {
  public abstract m mempty();
  public abstract NodeFactory<m> mappend();
}

abstract class IMonoidMaybe extends CMonoid<NMaybe> {
   public NMaybe mempty() {
    return new Pnothing();
  }
  public NodeFactory<NMaybe> mappend() {
    return PorElse.factory;
  }
}
```

When a type class member value is referenced, we would look up an instance in the environment for the inferred type.  This would yield the code for constructing the appropriate instance object, on which the referenced member can then be accessed.  

Functions with type constraints on type variables in their signature would be passed the appropriate instance object as an implicit parameter.  Looking up an instance in the environment for a type variable would result in code to access the parameter instance object.

Default implementations in type classes could be implemented simply as default implementations in the abstract class translation.

Type constraints in class and constructor signatures would simply instance objects passed as constructor parameters to the abstract class or concrete implementation class, respectively.

**Comment by ericvanwyk** [comment]
You have thought about this more than I have, but don't type classes naturally lend themselves to being Java interfaces in the implementation?  Where does that not work?

**Comment by krame505** [comment]
Yes, that was actually my first thought.  But consider how things work for translating a type class with multiple super-classes (e.g. `class Foo a, Bar a => Baz a`.)
These would become interfaces `CFoo`, `CBar` and `CBaz`, where `CBaz` extends `CFoo` and `CBar`.  Instance of `Foo` and `Bar` for `Integer` would become classes `IFooInteger` and `IBarInteger` that implement the appropriate interfaces.

But then how does the instance `Baz Integer` get translated?  It isn't possible for `IBazInteger` to extend both `IFooInteger` and `IBarInteger`, but it must still implement both `CFoo` and `CBar`.  This could be worked around by making `IBazInteger` dispatch to the appropriate functions in `IFooInteger` and `IBarInteger` explicitly, but this seems overly complex.  Also I think there might be issues with default implementations in this approach.

**Comment by tedinski** [comment]
> But consider how things work for translating a type class with multiple super-classes (e.g. class Foo a, Bar a => Baz a.)

I don't think you're making this confusion, but I wanted to spell it out because your example is easily misinterpreted. Type classes with "super classes" (is that the typeclass nomenclature for this? I forget.) need to be kept mentally distinct from implication instances or you'll go mad.

> `class Foo a, Bar a => Baz a`

is a type class declaration, so we just need a type: `interface Baz` with MAYBE an `extends Foo, Bar` -- not necessarily.

> `instance Foo a, Bar a => Baz a`

is an instance declaration, so it's a way of getting a value: `class UnnamedBazInstance implements Baz { Instance(Foo f, Bar b) { ... } ... }`

Which requires no multiple inheritance. In fact, there should be no inheritance whatsoever in the translation (not counting "interface extends" or "class implements", just "class extends").

You may or may not need to worry about dispatch functions (correspondingly, with or without that `extends Foo, Bar` bit back there). For one, I don't know what it'll look like when you dig into the translation for typeclasses, but I strongly suspect that while you'll want a way to get a plain `Foo` or `Bar` instance out of `Baz`, you might not need to use the same value at all. The de-sugaring might give you can explicit place to take a `Baz` dictionary and call `.getFoo()` on it as part of that process. I think this might be part of the search process that determines if a type class instance is inhabited.

But if I'm wrong, I don't think generating those dispatch functions is an overly complex solution. It's actually pretty core to why typeclasses are pretty nice whereas inheritance leads to all sorts of problems. (Also pretty core to, e.g., why forwarding is very different from inheritance.)

**Comment by krame505** [comment]
Hi Ted, thanks for your comments.

> I don't think you're making this confusion, but I wanted to spell it out because your example is easily misinterpreted. Type classes with "super classes" (is that the typeclass nomenclature for this? I forget.) need to be kept mentally distinct from implication instances or you'll go mad.

Yes, absolutely!  Yeah, that example was kinda brief.  

> You may or may not need to worry about dispatch functions...

I went through the same train of thought.  When I realized that the de-sugaring should make it easy to *not* go the dispatching route (as you mentioned), I was thinking that type classes would be implemented as abstract classes instead of interfaces because I imagined that `getFoo` and `getBar` would be a concrete methods in `Baz`, and `Baz` would have a constructor parameterized by `Foo` and `Bar` instances.  But guess this isn't actually necessary since `getFoo` and `getBar` can also be interface methods, with the instance objects stored in the concrete instance class.  

The other reason why I was thinking that type classes should be translated as abstract classes is that type classes can contain default implementations, which would be nicely implemented by non-abstract methods of the translated class.  However Java 8 now lets interfaces have default methods, so I guess this is a moot point since we already require Java 8.  

So I concede that type classes can be implemented as interfaces, and either the dispatching or non-dispatching approach is workable.  Although there is a certain elegance to having `class Foo a, Bar a => Baz a` translate as `interface Baz extends Foo, Bar`.
